types/wlr_xdg_shell_v6.c

```diff
@@ -166,23 +630,73 @@ static struct zxdg_shell_v6_interface xdg_shell_impl = {
 	.pong = xdg_shell_pong,
 };
 
-static void xdg_shell_destroy(struct wl_resource *resource) {
-	wl_list_remove(wl_resource_get_link(resource));
+static void wlr_xdg_client_v6_destroy(struct wl_resource *resource) {
+	struct wlr_xdg_client_v6 *client = wl_resource_get_user_data(resource);
+	struct wl_list *list = &client->surfaces;
+	struct wl_list *link, *tmp;
+
+	for (link = list->next, tmp = link->next;
+			link != list;
+			link = tmp, tmp = link->next) {
+		wl_list_remove(link);
+		wl_list_init(link);
```

Comment:
Any particular reason this isn't just wl_list_foreach?

Reply:
I lifted this from Weston, but it does look pretty much identical to `wl_list_for_each_safe()`.

Reply:
That seems to work ok.